### PR TITLE
Spreadsheet Interface for Student Data UI: Improve status message for the enroll process #8985

### LINKED ISSUE #8985: Spreadsheet Interface for Student Data UI: Improve status message for the enroll process
**Environment**:
`master` branch 

**Description of feature/enhancement**
Continuation of #8883. Similar to #8959 and should be done and merged first before bringing in features from phase 3 and 4 of the student data UI upgrade.

**Justification**
It provides a more user friendly approach for rectifying errors as users get direct feedback on what went wrong on the specific row that needs correction.

### PR DESCRIPTION
**Pull request by tanhengyeow**
Fixes #8985 

**Outline of Solution**
1. Created new AJAX action to verify enroll data.
2. The following error messages are now reflected upon AJAX instead:
    1. Empty enroll lines.
    2. Exceeded quota per enrollment (won't happen unless user updates the text area manually).
3. Enroll lines which have problematic fields after backend checks are highlighted in red. Error messages are shown in the form of a tooltip in the row containing the enroll line.
4. All cell styles are reset to white after a sort.
5. Enrollment would only be successful if all enroll lines are valid. This will trigger a submit to the current `InstructorCourseEnrollSaveAction`.

### CODE REVIEW

**Comment by tanhengyeow** [comment]
@damithc @wkurniawan07 @tran-tien-dat 

Ready for logic and code quality review before I modify existing tests/add new tests.

Changes are live in staging server.

File: src/main/java/teammates/common/util/Const.java [hunk 1]
```diff
@@ -953,6 +953,9 @@ private Const() {
         public static final String INSTRUCTOR_COURSE_STUDENT_DELETE_ALL = "/page/instructorCourseStudentDeleteAll";
         public static final String INSTRUCTOR_COURSE_STUDENT_LIST_DOWNLOAD = "/page/instructorCourseStudentListDownload";
         public static final String INSTRUCTOR_COURSE_ENROLL_PAGE = "/page/instructorCourseEnrollPage";
+        public static final String INSTRUCTOR_COURSE_ENROLL_AJAX_ENROLL_STATUS_PAGE =
+                "/page/instructorCourseEnrollAjaxEnrollStatusPage";
+        public static final String INSTRUCTOR_COURSE_ENROLL_AJAX_PAGE = "/page/instructorCourseEnrollAjaxPage";
```
Comment (tanhengyeow) [inline src/main/java/teammates/common/util/Const.java]: Might need to change the name of this action given that there might be more AJAX actions in the enroll page. 

Note to look into this after the current code behavior is acceptable.

**Comment by damithc** [comment]
> Changes are live in staging server.

Can you remind us the link to your staging server?

**Comment by damithc** [comment]
Some quick feedback:
* Users may not realize the error message is available on hover. We might need to pop up a dialog box to inform user of that fact.
* Related to the above, in both success and failure cases there should be a visible feedback e.g., dialog box. At the moment the page just stays as it is.
* Related to the above, may be the button should change to show a spinner while the request is being processed?
* It seems the section column is considered compulsory. It should not be.

**Comment by damithc** [comment]
Just realized that in the success case a new page loads. I had a situation where I tried to enroll a student, the line was highlighted in red, I fixed the error, and the line turned to white. i.e., no more errors shown but I remained in the enroll page and the student was not enrolled either.

**Comment by damithc** [comment]
image

Error in one line:
image

Update that line:
image

Everything is white now:
image

**Comment by tanhengyeow** [comment]
@damithc 

> Users may not realize the error message is available on hover. We might need to pop up a dialog box to inform user of that fact.

Would showing it as a status message be better? If the dialog box keeps popping up upon every enroll button click, it might be a little obstructive given that the user has to click `Ok` to acknowledge everytime.

> Related to the above, in both success and failure cases there should be a visible feedback e.g., dialog box. At the moment the page just stays as it is.

Same suggestion as above comment.

> Related to the above, may be the button should change to show a spinner while the request is being processed?

Sure, I'll look into this.

> It seems the section column is considered compulsory. It should not be.

Actually this is the current behavior since #8833. Since the row header is fixed, the backend always process the `Section` header passed along from the frontend.

 > Just realized that in the success case a new page loads. I had a situation where I tried to enroll a student, the line was highlighted in red, I fixed the error, and the line turned to white. i.e., no more errors shown but I remained in the enroll page and the student was not enrolled either.

I can't seem to replicate this in the staging server. With reference to the scenario pointed out, line 1 and 2 should be not acceptable as well since there are empty sections. 

I've tried several cases with the following steps and they all work fine:
1. Submit error rows, the error rows turned red
2. Correct only some rows, these rows become white and the error rows remain red
3. Correct all error rows, enroll result page is loaded

Does your course already have student entries with **no** sections and those rows in the picture are those existing students from that course? I have not played around with courses that have no sections yet since my development data only consists of students with sections, thus weren't able to catch this.

**Comment by tran-tien-dat** [comment]
From which commit does this PR start?

**Comment by tanhengyeow** [comment]
@tran-tien-dat Starts from https://github.com/TEAMMATES/teammates/pull/8989/commits/f44d7cf1642b8de1a4322beb3ef6ae9278155f04.

**Comment by tran-tien-dat** [comment]
Why do we need a separate action to verify if the submitted enroll data is valid? Whatever endpoint that is accepting enroll data should validate the data before actually carrying out the action anyway.

**Comment by tanhengyeow** [comment]
@tran-tien-dat 

The thought process of implementing this goes like this:
1. There's a choice between doing a page refresh, `createShowPageResult` or an AJAX request, `createAjaxResult` to show any error messages. After working on phase 3-4 of the proposed project, using AJAX and then updating the DOM immediately is preferred. This is because refreshing the page would cause the existing students' panel spreadsheet to be empty and when expanded upon click, does an initial AJAX request to load the latest existing copy of students' data. So in any case, `createAjaxResult` shall be used.

2. If I were to modify the current `InstructorCourseEnrollSaveAction`'s catch block where enroll exceptions occur to `createAjaxResult` instead of `createShowPageResult`, I can't find ways to capture the result back to the current enroll page and update the DOM, since the form will be already submitted. 

Therefore, I created this separate action to verify if the submitted enroll data is valid and if it is valid, utilize `InstructorCourseEnrollSaveAction` to do the enrollment of students.

**Comment by tanhengyeow** [comment]
@tran-tien-dat Upon a successful enroll, `InstructorCourseEnrollSaveAction` will perform `return createShowPageResult(Const.ViewURIs.INSTRUCTOR_COURSE_ENROLL_RESULT, pageData)`, which redirects to the enrollment result page.

If the whole action is converted to an AJAX action that returns an AJAX result, notably for the current enroll page to get any feedback data and update the DOM accordingly, how do we cater to the success scenario?

**Comment by tran-tien-dat** [comment]
We can just show a success message? What other info that the enroll success
page shows that needs to be displayed? We can find ways to show it here

**Comment by tanhengyeow** [comment]
@tran-tien-dat The enrollment result page is showing a different page, not the same enroll page we have been on. Example below:

image

**Comment by tran-tien-dat** [comment]
1. Just remove the result page and use ajax to display the necessary
feedback message on the current enroll page. I think we should either do
ajax all the way if you want flexibility or normal post form submission all
the way. It's awkward to do half this and that for the same logical action.

2. I wonder what's ghe difference between updated with no change and
unmodified 🤔

**Comment by tanhengyeow** [comment]
Sounds good, I also favor going for the AJAX path and removing this page. This would be more consistent with the behavior of phase 3-4 of the proposed project.

@damithc Do you have any objections or opinion on this?

**Comment by damithc** [comment]
> @damithc Do you have any objections or opinion on this?

No objections.

**Comment by tanhengyeow** [comment]
Closing this as the approach has shifted. Will be opening a new PR instead.

**Pull request closed without merging**